deps/jemalloc/src/jemalloc.c

```diff
@@ -3927,5 +3927,133 @@ jemalloc_postfork_child(void) {
 JEMALLOC_EXPORT int JEMALLOC_NOTHROW
 get_defrag_hint(void* ptr) {
 	assert(ptr != NULL);
-	return iget_defrag_hint(TSDN_NULL, ptr);
+    tsdn_t *tsdn = TSDN_NULL;
+    int defrag = 0;
+    rtree_ctx_t rtree_ctx_fallback;
+    rtree_ctx_t *rtree_ctx = tsdn_rtree_ctx(tsdn, &rtree_ctx_fallback);
+    szind_t szind;
+    bool is_slab;
+    rtree_szind_slab_read(tsdn, &extents_rtree, rtree_ctx, (uintptr_t)ptr, true, &szind, &is_slab);
+    if (likely(is_slab)) {
+        /* Small allocation. */
+        extent_t *slab = iealloc(tsdn, ptr);
+        arena_t *arena = extent_arena_get(slab);
+        szind_t binind = extent_szind_get(slab);
+        unsigned binshard = extent_binshard_get(slab);
+        bin_t *bin = &arena->bins[binind].bin_shards[binshard];
+        malloc_mutex_lock(tsdn, &bin->lock);
+        /* Don't bother moving allocations from the slab currently used for new allocations */
+        if (slab != bin->slabcur) {
+            if (bin->highest_slab_to_retain_inited && extent_snad_comp(slab, &bin->highest_slab_to_retain) > 0) {
+                defrag = 1;
+            }
+        }
+        malloc_mutex_unlock(tsdn, &bin->lock);
+    }
+    return defrag;
+}
+
+static long long ustime(void) {
+    struct timespec ts;
+    int res = clock_gettime(CLOCK_MONOTONIC, &ts);
+    assert(res == 0);
+    return (ts.tv_sec * 1000000) + (ts.tv_nsec / 1000);
+}
+
+static void init_defrag_bin_step(bin_t *bin, long long start_time, long long max_time) {
+    for (;bin->defrag_slabs_to_retain > 0; bin->defrag_slabs_to_retain--) {
+        extent_t *slab = extent_heap_remove_first(&bin->slabs_nonfull);
+        if (slab == NULL) {
+            bin->defrag_slabs_to_retain = 0;
+            break;
+        }
+
+        extent_heap_insert(&bin->slabs_nonfull_temp, slab);
+        if (!bin->highest_slab_to_retain_inited || extent_snad_comp(slab, &bin->highest_slab_to_retain) > 0) {
+            bin->highest_slab_to_retain = *slab;
+            bin->highest_slab_to_retain_inited = true;
+        }
+        if (ustime() - start_time >= max_time)
+            break;
+    }
+
+    if (bin->defrag_slabs_to_retain == 0) {
+        extent_heap_merge(&bin->slabs_nonfull, &bin->slabs_nonfull_temp);
+        bin->initing_defrag = false;
+    }
 }
+
+JEMALLOC_EXPORT void JEMALLOC_NOTHROW
+init_defrag(void) {
+    unsigned long long i;
+    tsd_t *tsd = tsd_fetch();
+    arena_t *arena = arena_choose(tsd, NULL);
+
+    for (i = 0; i < SC_NBINS; i++) {
+        unsigned binshard;
+        bin_t *bin = arena_bin_choose_lock(tsd_tsdn(tsd), arena, i, &binshard);
+        const bin_info_t *bin_info = &bin_infos[i];
+        uint64_t used_slabs = bin->stats.nonfull_slabs + (bin->slabcur ? 1 : 0);
+        uint64_t full_slabs = bin->stats.curslabs - bin->stats.nonfull_slabs - (bin->slabcur ? 1 : 0);
```

Comment:
Not sure if this is right, when `bin->slabcur` is not empty, it must be nonfull, and it will be counted in `bin->slabcur`, why `nonfull_slabs+1`?

Reply:
`bin->stats.nonfull_slabs` is supposed to be the size of the bin's `slabs_nonfull` heap. see: https://github.com/redis/redis/blob/60250f50c2f308d301fb0bfcb6d07faa2560ce60/deps/jemalloc/include/jemalloc/internal/bin_stats.h#L48-L49
So it isn't supposed to include `bin->slabscur` in the count.

Reply:
You are right.